{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5fd14047-9462-4e5e-9d4e-f1dff35bffb2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:17.248740Z",
     "iopub.status.busy": "2024-05-28T19:07:17.246472Z",
     "iopub.status.idle": "2024-05-28T19:07:23.392502Z",
     "shell.execute_reply": "2024-05-28T19:07:23.391018Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "here() starts at /scratch_isilon/groups/singlecell/shared/projects/Inflammation-PBMCs-Atlas\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "\n",
      "Attaching package: ‘limma’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘edgeR’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SingleCellExperiment’:\n",
      "\n",
      "    cpm\n",
      "\n",
      "\n",
      "Loading required package: iterators\n",
      "\n",
      "Loading required package: parallel\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(anndata)\n",
    "library(Matrix)\n",
    "\n",
    "\n",
    "library(SingleCellExperiment)\n",
    "\n",
    "library(here)\n",
    "\n",
    "library(edgeR)\n",
    "\n",
    "library(foreach)\n",
    "library(doParallel)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4fbac4f-d256-43b2-b71e-296654fb2abf",
   "metadata": {},
   "source": [
    "### Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d5a18dac-7260-446d-a184-0d799d43d35c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:23.419728Z",
     "iopub.status.busy": "2024-05-28T19:07:23.394509Z",
     "iopub.status.idle": "2024-05-28T19:07:23.429158Z",
     "shell.execute_reply": "2024-05-28T19:07:23.427908Z"
    }
   },
   "outputs": [],
   "source": [
    "adjPvalThr = 0.01\n",
    "FCthr = 1.5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f80c519e-7d13-4d49-84c9-8446bc45aed1",
   "metadata": {},
   "source": [
    "## Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "db846b7a-da31-4cde-8807-0b247a23345b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:23.432282Z",
     "iopub.status.busy": "2024-05-28T19:07:23.431221Z",
     "iopub.status.idle": "2024-05-28T19:07:23.440171Z",
     "shell.execute_reply": "2024-05-28T19:07:23.438972Z"
    }
   },
   "outputs": [],
   "source": [
    "adata_path <- here(\"03_downstream_analysis/02_gene_universe_definition/results/01_INFLAMMATION_annotated_pseudobulk.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "56b208f1-c11c-41c5-b5b6-cf6d43877bb8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:23.443213Z",
     "iopub.status.busy": "2024-05-28T19:07:23.442164Z",
     "iopub.status.idle": "2024-05-28T19:07:29.754045Z",
     "shell.execute_reply": "2024-05-28T19:07:29.752756Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 7357 × 14101\n",
       "    obs: 'studyID', 'sampleID', 'chemistry', 'disease', 'sex', 'binned_age', 'Level1', 'psbulk_n_cells', 'psbulk_counts', 'disease_cat'\n",
       "    var: 'hgnc_id', 'symbol', 'locus_group', 'HUGO_status'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "psdata <- read_h5ad(adata_path)\n",
    "psdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "be1651a4-164b-49bd-8247-f67ed106436a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:29.756893Z",
     "iopub.status.busy": "2024-05-28T19:07:29.755931Z",
     "iopub.status.idle": "2024-05-28T19:07:29.800507Z",
     "shell.execute_reply": "2024-05-28T19:07:29.799118Z"
    }
   },
   "outputs": [],
   "source": [
    "# makes sure that clinical covariates are available. Otherwise drop the patients\n",
    "psdata = psdata[!is.na(psdata$obs$binned_age)]\n",
    "psdata = psdata[!is.na(psdata$obs$sex)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb1a96de-ad75-45db-97d6-63bd895ade46",
   "metadata": {},
   "source": [
    "### Convert anndata to *SingleCellExperiment* format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "21d91543-a8a7-475c-8dc9-f065bfa192ab",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:29.808326Z",
     "iopub.status.busy": "2024-05-28T19:07:29.806951Z",
     "iopub.status.idle": "2024-05-28T19:07:31.958931Z",
     "shell.execute_reply": "2024-05-28T19:07:31.957629Z"
    }
   },
   "outputs": [],
   "source": [
    "# Generating the SingleExperimentObject. \n",
    "# Notice that the count matrix has to be transposed because in sce rows include genes, while columns include samples.\n",
    "sce <- SingleCellExperiment(list(counts=t(psdata$X)), \n",
    "                           colData = DataFrame(psdata$obs),\n",
    "                           rowData = DataFrame(psdata$var),\n",
    "                           metadata=list(experiment_info=\"inflammationAtlas\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "01b62261-08ed-4ccd-aa99-e03da7253f56",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:31.961599Z",
     "iopub.status.busy": "2024-05-28T19:07:31.960688Z",
     "iopub.status.idle": "2024-05-28T19:07:31.977663Z",
     "shell.execute_reply": "2024-05-28T19:07:31.976527Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>B</li><li>DC</li><li>ILC</li><li>Mono</li><li>Plasma</li><li>T_CD4_Naive</li><li>T_CD4_NonNaive</li><li>T_CD8_Naive</li><li>T_CD8_NonNaive</li><li>UTC</li><li>pDC</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'B'</li><li>'DC'</li><li>'ILC'</li><li>'Mono'</li><li>'Plasma'</li><li>'T_CD4_Naive'</li><li>'T_CD4_NonNaive'</li><li>'T_CD8_Naive'</li><li>'T_CD8_NonNaive'</li><li>'UTC'</li><li>'pDC'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item B\n",
       "\\item DC\n",
       "\\item ILC\n",
       "\\item Mono\n",
       "\\item Plasma\n",
       "\\item T\\_CD4\\_Naive\n",
       "\\item T\\_CD4\\_NonNaive\n",
       "\\item T\\_CD8\\_Naive\n",
       "\\item T\\_CD8\\_NonNaive\n",
       "\\item UTC\n",
       "\\item pDC\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'B'\n",
       "\\item 'DC'\n",
       "\\item 'ILC'\n",
       "\\item 'Mono'\n",
       "\\item 'Plasma'\n",
       "\\item 'T\\_CD4\\_Naive'\n",
       "\\item 'T\\_CD4\\_NonNaive'\n",
       "\\item 'T\\_CD8\\_Naive'\n",
       "\\item 'T\\_CD8\\_NonNaive'\n",
       "\\item 'UTC'\n",
       "\\item 'pDC'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. B\n",
       "2. DC\n",
       "3. ILC\n",
       "4. Mono\n",
       "5. Plasma\n",
       "6. T_CD4_Naive\n",
       "7. T_CD4_NonNaive\n",
       "8. T_CD8_Naive\n",
       "9. T_CD8_NonNaive\n",
       "10. UTC\n",
       "11. pDC\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'B'\n",
       "2. 'DC'\n",
       "3. 'ILC'\n",
       "4. 'Mono'\n",
       "5. 'Plasma'\n",
       "6. 'T_CD4_Naive'\n",
       "7. 'T_CD4_NonNaive'\n",
       "8. 'T_CD8_Naive'\n",
       "9. 'T_CD8_NonNaive'\n",
       "10. 'UTC'\n",
       "11. 'pDC'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] B              DC             ILC            Mono           Plasma        \n",
       " [6] T_CD4_Naive    T_CD4_NonNaive T_CD8_Naive    T_CD8_NonNaive UTC           \n",
       "[11] pDC           \n",
       "11 Levels: B DC ILC Mono Plasma T_CD4_Naive T_CD4_NonNaive ... pDC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(sce$Level1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b752b895-2cfc-4956-b828-bf805f2dd480",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:31.980300Z",
     "iopub.status.busy": "2024-05-28T19:07:31.979386Z",
     "iopub.status.idle": "2024-05-28T19:07:31.986672Z",
     "shell.execute_reply": "2024-05-28T19:07:31.985501Z"
    }
   },
   "outputs": [],
   "source": [
    "# merging togheter the two version of 5' sequencing protocol \n",
    "\n",
    "# colData(sce)$mainChem <- as.character(colData(sce)$ILCIC_chemistry)\n",
    "# colData(sce)$mainChem[startsWith(x = colData(sce)$mainChem, '5_prime')] <- '5_prime'\n",
    "# colData(sce)$mainChem <- as.factor(colData(sce)$mainChem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a1eef056-83a1-49d1-bdc6-bf57321188f2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:31.989248Z",
     "iopub.status.busy": "2024-05-28T19:07:31.988350Z",
     "iopub.status.idle": "2024-05-28T19:07:31.997518Z",
     "shell.execute_reply": "2024-05-28T19:07:31.996339Z"
    }
   },
   "outputs": [],
   "source": [
    "diseases <- unique(colData(sce)$disease)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ed14faa2-584f-4d67-bb08-2dff4310fb4b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:32.000110Z",
     "iopub.status.busy": "2024-05-28T19:07:31.999204Z",
     "iopub.status.idle": "2024-05-28T19:07:32.007436Z",
     "shell.execute_reply": "2024-05-28T19:07:32.006309Z"
    }
   },
   "outputs": [],
   "source": [
    "conditionPairs<- t(combn(as.character(diseases), 2, simplify = TRUE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3b4e13c2-6f77-452d-af46-a53baa1471e7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:32.010085Z",
     "iopub.status.busy": "2024-05-28T19:07:32.009182Z",
     "iopub.status.idle": "2024-05-28T19:07:32.017517Z",
     "shell.execute_reply": "2024-05-28T19:07:32.016402Z"
    }
   },
   "outputs": [],
   "source": [
    "conditionPairsFilt <- conditionPairs[(conditionPairs[,1] == 'healthy' | conditionPairs[,2] == 'healthy'),]\n",
    "#conditionPairsFilt <- conditionPairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fc37676-9659-40b9-8aba-54ec9076a5ac",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "**For each condition pair and for each cell-type:**\n",
    "\n",
    "subset the sce object to keep only a pair of conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0a8b6ae-aec1-4947-91a0-ba4750dc582f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:32.020149Z",
     "iopub.status.busy": "2024-05-28T19:07:32.019243Z",
     "iopub.status.idle": "2024-05-28T19:07:32.455452Z",
     "shell.execute_reply": "2024-05-28T19:07:32.454071Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "socket cluster with 19 nodes on host ‘localhost’\n"
     ]
    }
   ],
   "source": [
    "n.cores <- 19\n",
    "my.cluster <- parallel::makeCluster(\n",
    "  n.cores, \n",
    "  type = \"PSOCK\"\n",
    "  )\n",
    "\n",
    "#check cluster definition (optional)\n",
    "print(my.cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cf317cfc-6f40-4eeb-973e-ebec91685e97",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:32.458635Z",
     "iopub.status.busy": "2024-05-28T19:07:32.457590Z",
     "iopub.status.idle": "2024-05-28T19:07:32.473066Z",
     "shell.execute_reply": "2024-05-28T19:07:32.471840Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#register it to be used by %dopar%\n",
    "doParallel::registerDoParallel(cl = my.cluster)\n",
    "#check if it is registered (optional)\n",
    "foreach::getDoParRegistered()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cb278086-5621-47aa-a7ae-f1f791982664",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:32.476097Z",
     "iopub.status.busy": "2024-05-28T19:07:32.475073Z",
     "iopub.status.idle": "2024-05-28T19:07:32.483779Z",
     "shell.execute_reply": "2024-05-28T19:07:32.482556Z"
    }
   },
   "outputs": [],
   "source": [
    "annotationCol = 'Level1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "eb773599-5d63-467b-b1a3-28d071888fb6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:07:32.486755Z",
     "iopub.status.busy": "2024-05-28T19:07:32.485733Z",
     "iopub.status.idle": "2024-05-28T19:24:24.846816Z",
     "shell.execute_reply": "2024-05-28T19:24:24.845388Z"
    }
   },
   "outputs": [],
   "source": [
    "logFile <- here(paste0('03_downstream_analysis/02_gene_universe_definition/results/DEGlogFile_', format(Sys.time(), \"%s\"), '.txt'))\n",
    "TTfinal <- DataFrame()\n",
    "TTfinal <- foreach(cP_i = 1:dim(conditionPairsFilt)[1],\n",
    "                   .combine = 'rbind',\n",
    "                   .packages = c('edgeR', 'SingleCellExperiment')) %dopar% {\n",
    "# for(cP_i in 1:dim(conditionPairsFilt)[1]) {\n",
    "    diseaseA <- conditionPairsFilt[cP_i,1]\n",
    "    diseaseB <- conditionPairsFilt[cP_i,2]\n",
    "\n",
    "    # healthy vs. other\n",
    "    if(diseaseB == 'healthy') {\n",
    "        diseaseB = diseaseA\n",
    "        diseaseA = 'healthy'\n",
    "    }\n",
    "\n",
    "    \n",
    "    # subset sce object to keep only selected diseases \n",
    "    sce_conditionsAB <- sce[,(colData(sce)$disease == diseaseA | colData(sce)$disease == diseaseB)]\n",
    "\n",
    "    # remove celltype with less than 2 patients\n",
    "    TctA <- table(colData(sce_conditionsAB)[[annotationCol]][colData(sce_conditionsAB)$disease == diseaseA])\n",
    "    TctB <- table(colData(sce_conditionsAB)[[annotationCol]][colData(sce_conditionsAB)$disease == diseaseB])\n",
    "    \n",
    "    sce_condition_filt <- sce_conditionsAB[,((sce_conditionsAB[[annotationCol]] %in% names(TctA)[TctA >= 3] ) & \n",
    "                                             (sce_conditionsAB[[annotationCol]] %in% names(TctB)[TctB >= 3] ))]\n",
    "\n",
    "\n",
    "\n",
    "    # cat(file = logFile, append = TRUE, 'Cell types included: ', paste0(unique(sce_condition_filt$cell_mainGroup_21Sep23), join = ', '))\n",
    "\n",
    "    cellTypes <- unique(sce_condition_filt[[annotationCol]])\n",
    "\n",
    "    # cat(file = logFile, append = TRUE, sep = '',\n",
    "    #         diseaseA, ' vs. ', diseaseB,' -- Cell-type considered: ',paste0(unique(sce[[annotationCol]]), collapse = ', '),\n",
    "    #         '\\n')\n",
    "        \n",
    "    # for each cell types included\n",
    "    DEGfinal <- data.frame()\n",
    "    \n",
    "    for(cellType_i in cellTypes){\n",
    "        \n",
    "        sce_condition_ct_filt <- sce_condition_filt[,sce_condition_filt[[annotationCol]] == cellType_i]\n",
    "    \n",
    "        # create the edgeR object\n",
    "        y <- DGEList(counts(sce_condition_ct_filt), samples=colData(sce_condition_ct_filt))\n",
    "\n",
    "        # filter genes by expression\n",
    "        keep <- filterByExpr(y, group = sce_condition_ct_filt$disease)\n",
    "        y <- y[keep,]\n",
    "    \n",
    "        # cat(file = logFile, append = TRUE, 'Gene kept (removed): ',summary(keep)[3],' (',summary(keep)[2],')')\n",
    "\n",
    "        # normalize expression\n",
    "        y <- calcNormFactors(y, logratioTrim = 0.3)        \n",
    "        \n",
    "        # defining the design\n",
    "        # The idea is to correct for main sequencing protocol (i.e., 3'v2, 3'v3, and 5') and for sex (m,f)\n",
    "        # Unfortunatelly, not all the condition pair are sequenced with different protocols, \n",
    "        # so here we check if there are at least 2 different factor for each covariate to create an appropriet design      \n",
    "        desFormula <- '~0 + factor(disease) + factor(chemistry)'\n",
    "        # if(length(unique(y$samples$chemistry)) > 1){\n",
    "        #     desFormula <- paste(desFormula, '+ factor(chemistry)')\n",
    "        # }\n",
    "        if(length(unique(y$samples$sex)) > 1){\n",
    "            desFormula <- paste(desFormula, '+ factor(sex)')\n",
    "        }\n",
    "        if(length(unique(y$samples$binned_age)) > 1){\n",
    "            desFormula <- paste(desFormula, '+ factor(binned_age)')\n",
    "        }\n",
    "        \n",
    "        # cat(file = logFile, append = TRUE, 'design: ', desFormula)\n",
    "\n",
    "        cat(file = logFile, append = TRUE, sep = '',\n",
    "            diseaseA, ' vs. ', diseaseB,' ', cP_i,'/',dim(conditionPairsFilt)[1], \n",
    "            '. ', cellType_i,\n",
    "            '. Gene kept(removed): ',summary(keep)[3],' (',summary(keep)[2],')',\n",
    "            '. Design: ', desFormula, \n",
    "            '\\n')\n",
    "        \n",
    "        design <- model.matrix(object = formula(desFormula), y$samples)\n",
    "        \n",
    "        # estimate dispersion\n",
    "        # cat(file = logFile, append = TRUE, 'Estimating dispersion')\n",
    "\n",
    "        y <- tryCatch(y <- estimateDisp(y, design), \n",
    "                      error = function(e) { \n",
    "                        cat(file = logFile, append = TRUE, sep = '',\n",
    "                        diseaseA, ' vs. ', diseaseB,' ', cP_i,'/',dim(conditionPairsFilt)[1], \n",
    "                        '. ', cellType_i, ' FAILED!\\n')\n",
    "                          y <- NULL\n",
    "                      })\n",
    "        \n",
    "        if(is.null(y)) { \n",
    "            next \n",
    "        }     \n",
    "        \n",
    "        # fit a quasi-likelihood model\n",
    "        # cat(file = logFile, append = TRUE, 'Quasi-likelihood fitting')\n",
    "        fit <- tryCatch(fit <- glmQLFit(y, design, robust=TRUE), \n",
    "                      error = function(e) { \n",
    "                        cat(file = logFile, append = TRUE, sep = '',\n",
    "                        diseaseA, ' vs. ', diseaseB,' ', cP_i,'/',dim(conditionPairsFilt)[1], \n",
    "                        '. ', cellType_i, ' FAILED!\\n')\n",
    "                          fit <- NULL\n",
    "                      })\n",
    "        \n",
    "        if(is.null(fit)) { \n",
    "            next \n",
    "        }     \n",
    "\n",
    "        # defining contrasts (first and second column includes the two conditions\n",
    "        contrast <- numeric(length(colnames(fit$coefficients)))\n",
    "        contrast[which(endsWith(colnames(fit$coefficients),paste0(')',diseaseA)))] <- 1\n",
    "        contrast[which(endsWith(colnames(fit$coefficients),paste0(')',diseaseB)))] <- -1\n",
    "\n",
    "        # cat(file = logFile, append = TRUE, 'DEG testing')\n",
    "        res <- glmQLFTest(fit, contrast=contrast)       \n",
    "\n",
    "        # result extraction\n",
    "        DEGres <- summary(decideTests(res))\n",
    "        \n",
    "        nDEGs <- DEGres['Up',] + DEGres['Down',]\n",
    "\n",
    "        # select only DEGs\n",
    "        TT <- topTags(res, n = Inf)\n",
    "\n",
    "        # cat(file = logFile, append = TRUE, '# DEGs: ', nDEGs)\n",
    "        \n",
    "        # Adding comparison information\n",
    "        TT$table['conditions'] <- paste0(diseaseA,'_vs_',diseaseB)\n",
    "        TT$table['cellType'] <- cellType_i\n",
    "        TT$table['genes'] <- rownames(TT$table)\n",
    "        rownames(TT$table) <- NULL\n",
    "    \n",
    "        # merge results with previous data\n",
    "        DEGfinal <- rbind(DEGfinal, TT$table)\n",
    "    }\n",
    "\n",
    "    # TTfinal <- rbind(TTfinal, DEGfinal)\n",
    "    return(DEGfinal)\n",
    "    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "630acaa6-535f-4751-8c3f-e32fc4b4561a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:24:24.849509Z",
     "iopub.status.busy": "2024-05-28T19:24:24.848583Z",
     "iopub.status.idle": "2024-05-28T19:24:24.871658Z",
     "shell.execute_reply": "2024-05-28T19:24:24.869914Z"
    }
   },
   "outputs": [],
   "source": [
    "parallel::stopCluster(cl = my.cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "52a70a69-a213-4995-bd92-2bf42c358831",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:24:24.874587Z",
     "iopub.status.busy": "2024-05-28T19:24:24.873624Z",
     "iopub.status.idle": "2024-05-28T19:24:25.089686Z",
     "shell.execute_reply": "2024-05-28T19:24:25.088307Z"
    }
   },
   "outputs": [],
   "source": [
    "TTfinal$FDRfinal <- p.adjust(p = TTfinal$PValue, method = 'BH')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2ac0c81f-11d4-40fc-9d90-dcf20b4a982f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:24:25.092308Z",
     "iopub.status.busy": "2024-05-28T19:24:25.091406Z",
     "iopub.status.idle": "2024-05-28T19:24:25.147905Z",
     "shell.execute_reply": "2024-05-28T19:24:25.146515Z"
    }
   },
   "outputs": [],
   "source": [
    "DEG <- unique(TTfinal$genes[TTfinal$FDRfinal < adjPvalThr & abs(TTfinal$logFC) > FCthr])\n",
    "disease <- TTfinal$conditions[TTfinal$FDRfinal < adjPvalThr & abs(TTfinal$logFC) > FCthr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a8a8434b-9901-41bb-aa21-8ead65d6681b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:24:25.150529Z",
     "iopub.status.busy": "2024-05-28T19:24:25.149626Z",
     "iopub.status.idle": "2024-05-28T19:24:25.159483Z",
     "shell.execute_reply": "2024-05-28T19:24:25.158406Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "6868"
      ],
      "text/latex": [
       "6868"
      ],
      "text/markdown": [
       "6868"
      ],
      "text/plain": [
       "[1] 6868"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(DEG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e84f23ab-dede-424f-a445-927e11663105",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:24:25.162027Z",
     "iopub.status.busy": "2024-05-28T19:24:25.161136Z",
     "iopub.status.idle": "2024-05-28T19:24:25.170980Z",
     "shell.execute_reply": "2024-05-28T19:24:25.169905Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "19"
      ],
      "text/latex": [
       "19"
      ],
      "text/markdown": [
       "19"
      ],
      "text/plain": [
       "[1] 19"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(disease))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "700f3446-025a-4392-b8f3-f73e2bff9d41",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:24:25.173515Z",
     "iopub.status.busy": "2024-05-28T19:24:25.172625Z",
     "iopub.status.idle": "2024-05-28T19:24:25.221685Z",
     "shell.execute_reply": "2024-05-28T19:24:25.220322Z"
    }
   },
   "outputs": [],
   "source": [
    "TTfinal_DEGfilt <- TTfinal[TTfinal$FDRfinal < adjPvalThr & abs(TTfinal$logFC) > FCthr,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b06d0dc8-fc50-40ed-938b-5979b31c698c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T19:24:25.224288Z",
     "iopub.status.busy": "2024-05-28T19:24:25.223388Z",
     "iopub.status.idle": "2024-05-28T19:24:28.776092Z",
     "shell.execute_reply": "2024-05-28T19:24:28.774720Z"
    }
   },
   "outputs": [],
   "source": [
    "write.table(x = TTfinal_DEGfilt, file = here('03_downstream_analysis/02_gene_universe_definition/results/03_DEhealthyVSothers_DEGsel.csv'), sep = '\\t', col.names = TRUE, row.names = FALSE, quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "25de5fc9-f7f4-4375-ae9d-8ebd79226693",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T17:47:23.140396Z",
     "iopub.status.busy": "2024-05-28T17:47:23.139435Z",
     "iopub.status.idle": "2024-05-28T17:47:28.258645Z",
     "shell.execute_reply": "2024-05-28T17:47:28.257127Z"
    }
   },
   "outputs": [],
   "source": [
    "saveRDS(object = TTfinal, file = here('03_downstream_analysis/02_gene_universe_definition/results/03_DEhealthyVSothers_allGene.rds'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "7a28036c-995a-4a2f-b405-e2cc45dc82e9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T17:47:28.261588Z",
     "iopub.status.busy": "2024-05-28T17:47:28.260568Z",
     "iopub.status.idle": "2024-05-28T17:47:42.375928Z",
     "shell.execute_reply": "2024-05-28T17:47:42.374420Z"
    }
   },
   "outputs": [],
   "source": [
    "write.table(x = TTfinal, file = here('03_downstream_analysis/02_gene_universe_definition/results/03_DEhealthyVSothers_allGene.csv'), sep = '\\t', col.names = TRUE, row.names = FALSE, quote = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01925abe-1f5e-41d6-ac5a-3dd5e0c2d042",
   "metadata": {},
   "source": [
    "### Exploring list of DEG - Healthy vs. each other condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "90901d25-e259-49d9-b738-244e61ab9b91",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T17:47:42.378844Z",
     "iopub.status.busy": "2024-05-28T17:47:42.377836Z",
     "iopub.status.idle": "2024-05-28T17:47:44.173047Z",
     "shell.execute_reply": "2024-05-28T17:47:44.171555Z"
    }
   },
   "outputs": [],
   "source": [
    "# Loading the results\n",
    "TTfinal <-  readRDS(here('03_downstream_analysis/02_gene_universe_definition/results/03_DEhealthyVSothers_allGene.rds'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "77e3dfe6-b062-4e58-9f0c-aabe83c7c313",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T17:47:44.175992Z",
     "iopub.status.busy": "2024-05-28T17:47:44.174972Z",
     "iopub.status.idle": "2024-05-28T17:47:44.213671Z",
     "shell.execute_reply": "2024-05-28T17:47:44.212444Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "6868"
      ],
      "text/latex": [
       "6868"
      ],
      "text/markdown": [
       "6868"
      ],
      "text/plain": [
       "[1] 6868"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DEG <- unique(TTfinal$genes[TTfinal$FDRfinal < adjPvalThr & abs(TTfinal$logFC) > FCthr])\n",
    "length(DEG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "3a547eee-f29a-48df-8f70-6e3b16c6001c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T17:47:44.216428Z",
     "iopub.status.busy": "2024-05-28T17:47:44.215421Z",
     "iopub.status.idle": "2024-05-28T17:47:44.253624Z",
     "shell.execute_reply": "2024-05-28T17:47:44.252344Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "   healthy_vs_asthma      healthy_vs_BRCA        healthy_vs_CD \n",
       "                1377                 6592                 1368 \n",
       "healthy_vs_cirrhosis      healthy_vs_COPD     healthy_vs_COVID \n",
       "                 906                  739                 1614 \n",
       "      healthy_vs_CRC       healthy_vs_flu       healthy_vs_HBV \n",
       "                2985                 2041                 7510 \n",
       "      healthy_vs_HIV     healthy_vs_HNSCC        healthy_vs_MS \n",
       "                4394                 1532                 4032 \n",
       "      healthy_vs_NPC        healthy_vs_PS       healthy_vs_PSA \n",
       "                6602                 1624                 1450 \n",
       "       healthy_vs_RA    healthy_vs_sepsis       healthy_vs_SLE \n",
       "                1425                 1208                  210 \n",
       "       healthy_vs_UC \n",
       "                1541 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(TTfinal$conditions[TTfinal$FDRfinal < adjPvalThr & abs(TTfinal$logFC) > FCthr])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "aef8e7ed-34b1-4ba1-a07a-a299d40f2aeb",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T17:47:44.256485Z",
     "iopub.status.busy": "2024-05-28T17:47:44.255492Z",
     "iopub.status.idle": "2024-05-28T17:47:44.363204Z",
     "shell.execute_reply": "2024-05-28T17:47:44.362104Z"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAADAFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGC\ngoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OU\nlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWm\npqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4\nuLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnK\nysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u\n7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////i\nsF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3deWAU5f3H8W8uEkBuFRQQFLzQ\nelUUsYpSFa1nPaqi1rMi1Uq9StUi3ngU8b7iCVZAqlSBKnjEnxdUURQUD+QS5QoKKMgRsvPb\n2SsbCDtx8tlMknm//sg+u/vszLDJm+xunmzMAVBjFvQBAA0BIQEChAQIEBIgQEiAACEBAoQE\nCBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQ\nEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIg\nQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBI\ngAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAA\nIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEB\nAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKE\nBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgEGdLCV728vj7AwwOq\nL8iQzi9o7iFnfICHB1RfkCGdc/wMD1v8J8DDA6qPkAABQgIECAkQICRAgJAAAUICBAgJECAk\nQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBGoQ0uQre+/aadfeV07xuwFCQoPh\nO6SVx1nScSv9bYKQ0GD4Dukw2/veyfMWz5t8z552uL9NEBIaDL8hvWt/LE8MN5xpk31tg5DQ\nYPgN6S6bkxrPtrt9bYOQ0GD4Dek2W5Yal9rtvrZBSGgw/IY01u5JjYfZi762QUhoMPyGtGa7\n/IGzYqOvrszrtMbXNggJDYbvV+0+2MqsZbe9u7U022qqv00QEhoM/z+QXXbNzjlmlrPzNcu8\nJ1eJkNBg1GiJ0KpZ02et8n9zQkKDwVo7QICQAAEWrQICLFoFBFi0CgiwaBUQYNEqIMCiVUCA\nRauAAItWAQEWrQICLFoFBLK0aHXa1Aqvbe7mhIQGIztr7WbnW5qyzcwiJDQY2V+0+q6t28w1\nhIQGI/uLVgkJIZD9RauEhBDI/qJVQkIIZH/RKiEhBLK/aJWQEALZX7RKSAiB7C9aJSSEQPYX\nrRISQiD7i1YJCSGQ/UWrhIQQyP47rRISQoC1doBAjUNauWBD5gmEhBCocUiDbUHmCYSEEPAb\nUnlZwiCbV1ZWnmEmISEE/IY0OP0392xwhpmEhBDwH1KzFjGF1rxFiyEZZhISQsD3EqHC7jNj\nA54jATV4sWH6nkV3uk+NCAmoyat26wbm9vyKkABXTV7+fnv7JvdGCAmo4c+RfjzfDj2HkICa\n/kD2xa2NkIAar2xYesO1Hn/4kpAQAixaBQQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJ\nECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAg\nJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRA\ngJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQ\nAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAAB\nQgIECAkQICRAgJAAAUICBAgJEKhBSGvG39j/7P43jl+TeRohIQT8h3RXS4treVfGeYSEEPAd\n0pVWdOKQp0Y9NeT3hXZVpomEhBDwG9KsnH2/Swy/3Sf36wwzCQkh4Dekh2xqavyBPZxhJiEh\nBPyGdKutSI2X260ZZhISQsBvSE/bhNR4vA3PMJOQEAJ+Q1rapN3LieH4tk1LM8wkJISA71ft\nnsy1Dif8+bI/H9/ecp/KNJGQEAL+f45U0jMn9mOknJ4lGecREkKgJkuEFk4ovrd4wkKPWYSE\nEGCtHSBASIAAi1YBARatAgIsWgUEWLQKCLBoFRBg0SogwKJVQIBFq4AAi1YBARatAgIsWgUE\nsrPWbt42rVKa2drNzCIkNBjZCals7HMpN/EdCQ0fi1YBARatAgIsWgUEWLQKCLBoFRBg0Sog\nwKJVQIBFq4AAi1YBARatAgIsWgUEeINIQICQAIEahBRxP5S/PWpa5mmEhBDwHdKPZxU1H7Bh\neXczO+KnTBMJCSHgO6S+lpNrN5yXd/hJbeyPmSYSEkLAb0hz8/ZYuLxnu2bjHGfBdnmLMswk\nJISA35CetZGOM8EOdcfD7LkMMwkJIeA3pDvsY8eZbX91x2/ZnRlmEhJCwG9IQ2264yyywe74\nU1Z/I+z8hvQve81xlh/9L3c8yZ7IMJOQEAJ+Q/rMhqbG99o7GWYSEkLAb0iRQ65LjffbYn2G\nmYSEEBAsEYpMmZ7pakJCCLDWDhAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFC\nAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIE\nCAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJ\nENgopLKf5HsgJIRAWkif3dS7dY4V7Xj+qDXKPRASQiAV0rheZtZi+13bFpi1uapUtwdCQggk\nQvr8t9bqohcWucP179/dw1r8U7YHQkIIJEIa3OXptWmXfnlunmwPhIQQSIQ0p2yjy7+U7YGQ\nEAK8/A0IEBIgkB7S159kYw+EhBBID+mM3aIfflgs3gMhIQSSIa2IJEKKfVAiJIRAMqTBW536\naB9CAvxJhjS5b1uzxn3v/+g0QgJ+sbTnSJ/0LGptZq3v+0q6B0JCCFR+sSEy87FuOWbbjxHu\ngZAQAsmQnh389vrEiw2zHjhOt9KOkBAKyZCGmjVt3f6dtbzYAPiQemi3ZFS/rcwKf9O1q/S3\nkQgJoVDpOdIOIy7aI9eKjhg6X7gHQkIIbLKy4YxtB+xqg4V7ICSEwKYhRT98M1O4B0JCCKSH\nNPllh5UNgB+b/BrFhCfFeyAkhAC/jwQIEBIgkAhp9caXb3KBb4SEEEiEdP3vZ6RfuvbedrI9\nEBJCIBFSSXs76LFv4uOf3+zfOud02R4ICSGQfI60enBrs20PObHvkXvlmx3ynm4PhIQQqHix\n4ecnjm5hUTm7DZiR4Qa/GCEhBCq9ardh9uQ3pv8g3gMhIQR4+RsQICRAgJAAAUICBAgJECAk\nQKAmIa14a/STo99a4TGLkBAC/kP6X5889+e3ln/k+xnnERJCIBXSnJted5zSLq6d5lTjhqPy\nbcs+51x0Tp82lj8600RCQgikQjqjaJ7jLIp9j7EzvW/3fbMtX4jERuX/btPs+wwzCQkhkAzp\nx6I/Om5IVyxatOjkopWetxtuL6bGY+2ZDDMJCSGQDOkVc9/ve1HsjbhesEmet7vVKl5kWG5D\nMswkJIRAMqR7zH0LrnhIH9t9nrd7wKalxh/aQxlmEhJCIBnSzbYg+nHt2M+jH7+xmz1v90XO\nAUsTwyX753yZYSYhIQSSIQ2zT1OXzbB7vG94sW1xxt3PvfjcsL5N7ZJMEwkJIZAMaaw9lbrs\nSavG12/5DY3jL/FZ4xsjmSYSEkIgGVJp/gHJHCIH5C+rzk1XjLzqjBPPuGqkx0t8hIQQSP0c\nqa9dXh4blF9WnZ8jVR8hIQRSIS1qZ7vfMWnypDt2t3aLlHsgJIRAxVq7L7slnvPsluk1uHQs\nWgUS0hatbhjZd5+u+5wxckP1bsmiVSDF9+pvFq0CFfyGxKJVIE0ypEXTKr7cV434xPN2LFoF\n0iRDGmxzHWdep2HueG41/oYsi1aBNJVCmhUvqDohsWgVSOM3JBatAmn8hsSiVSCN75BYtApU\n8B0Si1aBCjUIKYPll1yYcjwhoeGrCKlZixbNrLBFC/dkcA23uuzMU1J6ExIavoqQ0g2u1m1Z\ntAokJEMqW5OurBq3ZNEqkMKiVUCgUkg/z5o6a031bseiVSBNWkivHNoo+kit8LDXq3M7Fq0C\naSpCui5aUatOLaMfb6nG7Vi0CqRJhTTBim7+Nnr6zY2F9qr37Vi0CqRJhdQjryQxejX3IO/b\nsWgVSJMM6Vvrm7rsNFvofUMWrQIVkiG9FPtrFHHP2XjvG7JoFaiQDOn+2F+jiPvUHqjOTVm0\nCiQlQ7rVvk1d9m3GV+F+KUJCCFSstat4e9VFNV60mo6QEAKEBAhUhNS8RVLzaoU0e9B5d8b/\nasXiPiMyzCMkhIDvX6N4w33RrtVL7jDzLwISEkLA769RrO9oJxdf3jjX/fNkhITQ8/trFBPt\nrOjHL7rm/ouQAN8hDbUS92RJt7wxhAT4DWmIfRY7Xdi1YBwhIfT8hvSMTYwP5ncsfJSQEHZ+\nQ/rUBiZGX7bNIySEne/3bNixw/rEaHprQkLY+Q5p0rDvksOpAzP9JiAhIQR8h1RthIQQICRA\ngJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQ\nAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAAB\nQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUIC\nBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQI\nCRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECgRiGVLZm7pMxr\nEiEhBPyH9OOQfRuZWeG+t/2UcR4hIQR8hzS9vVlR5106F5m1n55pIiEhBPyGtLZL7qXTI9FB\n5JNLcrtuLhUXISEE/Ib0kv0zNb7DxmWYSUgIAb8h3W6lqfFSuyPDTEJCCPgNaajNS43n2rAM\nMwkJIeA3pLftgkhiGDnP3skwk5AQAn5DihxsPYo/Xrpy6bRH97ODIxlmEhJCwPfL38t6W1Lv\nZZkmEhJCoAYrGyb1696+dfvu/SZlnlaTkPK7/NrDUf6PH9Cp22vtck4bnNkfc7N+/EA11PGQ\n7vWY8BghoU6o24tWCQn1RN1etEpIqCfq9qJVQkI9UbcXrRIS6om6vWiVkFBP1O1Fq4SEeqJu\nL1olJNQTdXvRKiGhnqjbi1YJCfVE3V60SkioJ+r2olVCQj2RnbV2kTdfTbmbkNDwZSek2Y0t\nzdrNzCIkNBgsWgUEWLQKCLBoFRBg0SogwKJVQIBFq4AAi1YBARatAgIsWgUEWLQKCLBoFRDg\nDSIBAUICBAgJEFCEVHr22AzXEhJCQBHSXBuc4VpCQggQEiDgN6QPHqpwMyEh7PyGNDj9d8kJ\nCWHnN6Sb7ZQrkv5ESAg7vyE9b4+kxjxHQuj5DWmOXZQaExJCz/erdq32Tw0XHfh4homEhBDw\nHdLq1dWcSEgIAZYIAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKE\nBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQI\nEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAjU85AGm6d+Wf8XAvU9pCtzR3s48qSs/wuB+h+S\n1yb6EhJqASEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBAS\nIEBIgAAhAQKEBAgQEiBASIAAIQECDT6kfRq18tCmJOv3ARq8Bh/STnsM9dByRNbvAzR4DT+k\nw71mtCUk1BghERIECImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCACEREgQIiZAgQEiE\nBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASDMKm3q9PcrOZVm/l1DPEdKMvNM9\n3h3lr/ZT1u8l1HOENCPvVo8JowgJXgiJkCBASIQEAUIiJAgQEiFBgJC8Q3rUDj3Mw51ZvxtR\ntxGSd0i32iWXZbbf4Vm/G1G3EVJ1Qvqfx4wLCSnsCImQIEBIipBO7jzQw9Vzsn5HI0iEpAhp\n7xY9PDR+MOt3NIJESJKQDvDaSZcHsn5HI0iEREgQICRCggAh1U5Ijc1LkxVZ/1QgewipdkJq\ndP7ozIbZwqx/KpA9hFRLIV3rMWEcIdVrNQqpbMncJZ6/hU1ILs+QnrczL/QwvCafK2SX/5B+\nHLJvo+hD+8J9b8u8NpqQXJ4hPWyHHJHZ9js84uFRvqkFxndI09ubFXXepXORWfvpmSYSkqsa\nIb3hMeM3jTp4aHSH388maspvSGu75F46PRIdRD65JLfr5lJxEZJLENKBe3jtZJcLXvXwsNeE\nV7/0+QURdn5Desn+mRrfYeMyzCQkV62E1NTzNXZvjXbwsM0fbvNw8s1eM576wcsGn1+XFco/\nmurlhxrvpILfkG630tR4qW38kGLOVhVvCtfM1m9mG+cXNPdgTTwmFJnXJvK8d9LYY0ITa+a1\nk3zPnRR57mQLjxn5eV47ySn0mNDUcycFghYVirzeazCvuceELbx30t/nF39V/IY01OalxnNt\n2EbXlpdUPFaY9MzmtrHQ83HGoy97TJj4iNcmxjzrNeOxCR4TJj3ktYkXRnjNeGKc14wHvSb8\n52mvGU+/WOOdjHvCa8aIF7xmPDTJY8KEx7w28ewYrxmPTPSY8PKjXpt4VfnajN+Q3rYLIolh\n5Dx7R3U4QP3kN6TIwdaj+OOlK5dOe3Q/OzjifQOgIfP98vey3qmHmr2XKY8IqIdqsLJhUr/u\n7Vu3795vku5ogHoq+2vtgBAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQI\nCRAgJECAkAABQgIECAkQCDKkHll8OyfUb/Xur1sHGVLfYz3fwq82tL0x6CNwPW+vBH0IroE7\nBH0EMb37Bfh16UuQIZ1zToA7r9CxTvyVhy/qxp91eaBb0EcQc9Jfgj6CX4qQCCkdIflESISU\njpB8IiRCSkdIPhESIaUjJJ8IiZDSEZJPhERI6QjJJ0IipHSE5BMhEVI6QvKJkAgpHSH5FGRI\nF14Y4M4rdBkV9BG45uSUek/KvuK9gj6CmNOvCPoIfqkgQ/pB+cdw/ZtfFvQRxMwO+gBi1i0I\n+ghiSlcGfQS/FL9GAQgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQ\nEiBASIBAgCF9/7+P1wa3d2fD3PdmbkidWzPt/cB+O2p2SUkkMQzwPpk3eebPyfE3730RyTQ3\na9bOmjI/9UmJfD65bvx2VHUEFtKyU/PMmt8czOfLcV45uamZtRq0LnaufNAWZvlnrgjkUL7f\nxiz+u4UB3ifPdI3eHYUXxcYf7xcdd3qp9g9i+aWNo3tud1s8pTEdomcOnFn7h+FLUCGt724n\njHm8m/0joP0f3+h3N90/oLn9IXZugO355HNH28HlQRzKKe0ax0MK8D75u7W9/IEh557qjme3\nzrty7J2t8l6p7YOI9LQDHxt7e3u73D33Qs6WQ8cOyG37XW0fhj9BhfSgHR/9+H27gjnB7H9c\n7B0SZjW3j6Inn+Z2XBn9PPaxZwI4khH2Uot4SMHdJ+Ntv4rvxn+wYdGP79qOtf2d8SPbcX30\nZKY1j34sa587NXpyq11Qy0fhU1AhHWCT3ZPr7JaADiDuZBsR/Xi1DXHPvBHEn7f6puU5TiKk\n4O6T/W1GavxjYfPY07SD7d1aPoo3rY97sj6/VfTjxPiZVY2bBflEuvoCCml9o8axB8Kv27HB\nHEBCH5sU/XhI/GtmXV7zWj+AyG87rEiEFNx9sth2/37QoQf3/9A982biv5NBNrSWD2NFy6Lo\nw8n119pV0TPXJ/47Ocg+rOXD8CegkObajrHTr233YA4g7sO8zu6rDZ0s/kC8gy2r7SMYZhOd\nREjB3SeTrGfHgl9tb7kPRc88aX+KXfiYXVLLh+G83No67tVquzvdp6pn279il51p/67tw/Al\noJBm2N6x08XWOZgDiFnUOe8N97SNxd/9aRebV8tHMLPI/RuP8ZCCu09Gme3yteO81ChvpuPc\nZ5clLqz19+/c8GDzTt1bN7lwdXR8kr0Yu+wie6q2D8OXgEL6wn4VO11gOwVzAK7Fu+U+ERu0\ns+9jp11q+91O1+/T+ScnGVJw98nzFnuE6/zFrnachy3+LqcjrLbfv3PDYUX/jd4n/4g9OzrN\nno9deL49W8uH4U9AIS2zdrHTaXZAMAcQtWCn3MR/drvbF7HTFrlravcQ7rHrS6Ka2mslawO8\nT/7PzP0m4Iy23zvOv+202IXD3Kpq1eN2Zex0T5vmOJfYw7Ezx7uPfeuBoF61a2dL3ZPhiQfk\nAfi6U/7IxPA0G+OefJN4klJ7rrWUBQHeJ9/n2Cr3dJSd5H5j3C124XmJJym150J7LHYa+3Q8\nbBfHzuxg9eMHSUGFdJ65T2yd42xsQAfw6TaNUrsebrGfRN5lA2r5ID4eEdPEnh6xOsj75EB7\n2T252K533Ae47mqCtVvn1/abkV9mf3VPIrvbW44zP6eD+0Olj2zPWj4Kn4IK6bP8LaPfv4ut\nW0BvvD21TeOXU2fWbp8z3HHeb1kY0E+HEz9HCu4+edl2ij64fb6g8fzomUesR6mz7txaf4rk\nvGmFL0Q/GVdZR/el1NOt/3pn8T715ClScGvtHsvP3X07a/tZQLvvaU06xTzonvuwjXXeLacg\nqM9ZIqQA75Obc3J32tYax57fR86zxnu1th4/1fpRXJtjbfdoZq3fcc/8sLdtuVdR7b8G71Nw\nq7+nDzjyuFtq/ec2Sf17JcSfCCy56dijLg9sfeTRvRIrnoO7T97/y5HH/GNu4syEc4849fEg\nHipM/dsxvz1laOIh5bpHTjnivEkBHIUv/D4SIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIg\nQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBI\ngAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEgC5SVRb81Yu9HFC+z4DDfKfK3YL95ZZP+9I8nx\nlE8c57uSn93hDy36aQ+swSAkgTUW0/hPKypd/G3eiRlulPlasV+8s+E2ITlcXXCp41zcOP7f\nxI15M6QH1mAQksAaazJy5PDr2tmhQR+JSmTHHVPjifaS4+x8RPxMacHpAR1SHUdIAmushXsy\nt6lNdZwN86ZMiz0MctaWuP97/1TyuVP+2f8tKfnUvezdkh+iH0tL5ieujd74y/e+WB3fTvnn\n734ZqbTlSPTKxCPG1R9Oif+VYneDZZ986O6jfOaUH5MXrf/k/VWJW1UcQmLnK6vYWaXtbZgx\nZXn6fifazanxwLyV0YeGtyfOndBoia/7qKEjJIFESM5BNtLp2zr6IK/wz+7XcfyJyTQ7+9/b\nmL3drFv0TGmO3Rs9ucVeSFy77m/NovMLTooOI7dvGR12GF6x3fI72kYv2eLq6HBVv0KznD5z\nEhvc2qzN885r20cfTz4Sv2hEm+j4htjN0g4hsfPJm+ys8vbGd4he/Pe0hs+0WPbO7OLi4l23\nLS7ub1cWFy91L3nK7s/eHVmPEZJAMqSuNs7pMvDpV4YfbX2dipC6FVzw+L2zj7aFjvOc5R8X\nvax37g+Ja6+1w0a+PmrQ0dHhuXbM0y8/uIs9mdruWbbLAxNHDIh+4W841A4bPf5Pts1id4O7\nNh045qaigjGNLx19Z/O8We5FXZve8NqIbjbYvVnaISR3vvHOKm+vW5PLRt+3TdqOnQ7Ny2On\nIy3NZPeSL60Wn9rVI4QkEA9pzd9ti+Tjo5Nz5laEZE+7F91lIxynX5MzWmxw1hT9Onnt3kXr\nEjd5xa5yT5Z3bFueuGScdUs8DHOesYPcS6+wC2Mb/G90fLu5G3SetitjF7nj71sVLEwdVOwQ\nkjvfeGcbbW9cdPyBHZi67bfWMz5YV1o63EaXlm57ZGlpaVnsosbb1vwOa4AISWCN5e688w6F\n1ujZ6Jn1U/8zauQF0Qd5qZB2iD1m+tjOcZwdj3jGpjivu83Erz085/XERvra/WNcB9vMxCWn\nRB8AJhxvr7gnywraRKIb3NMdf2gd3ZPv7AR3Hx1j9f3N3Ad6aYeQ3PnGO6tie07bVql/0FQ7\nLjUemLs8upNbU+c75FV+GocYQhJYY3m9evU+8YbZ0fFDbeKPg+6rCOmY2KTIVh2db+z2hXaL\nc437ZRy/9rUi2+3ike43st1SD6HeSWx2V/s2uYddbHHsdAdbEt3gqe7wG/ute7LeDnD3cVTs\n6mfsssqHkNz5xjurYnvOHvmpf9BrdkZq/JtfOc4YezN1vputlNxpDQwhCSSfIznul9xe//1u\n1ZqHbVjaiw3xq06xr560qc6uvZ39C1alfkb69dW/zrcmdzvO9rakLC65re0t+TKc09nir8Lt\nbfOTG1xg7hMrp8z2d/fxh9jVL1r/yoeQ3PnGO6tie86eeal/0FQ7NnY6vbi4uOiA4uJj7J7i\n4sTDTL4jVYmQBNJCOtY+d0+urSKkh+2hs1qVO5cULs4/yElfbPDTqHY505we9sZGm+1p05PD\n/e2L2Gnr3NVVh3RA7Or73Vcb0g9ho5BSO6tie+khfWc9Yqd3pr/WsCB2UYTnSFUiJIG0kHqZ\nu7qhbKcqQvrKTm5/ouO8YNN/2LMAAAJfSURBVJfa9U7lVTu32aPOoMSjsArX2YDk8Aq70T2Z\nZPs6VYdUMN89c6i9VvkQNg0pvrMqtpcektNhi9hzrp9LS6/KnVX6ec41paWl8ddAvrCTfvkd\nFAKEJJAW0l9sUMRZd3FeFSE52xXaA46zPLfQ3nKS1w52v+msOtRedZZuZYPXuuv2rkhua0mb\n/CeiD6OWjY8+JitqHr3N/F1t1GZCyjnqR/d73q8ilQ+hckgVO6tie5VCOtM+SYx6dnecZ+1/\nqWv4OVLVCEkgLaSFW1ung9psXdVDO+cciz2g2teauq9Cx69tY1t379Es9tx+akdr3qN7C9sj\ntd33tra2PXcu6BUdjm5ku3UvcF/qrjqkk/Zs8ZsutvVnGx1C5ZDSdrbp9iqFNNHiP9t1VuRd\n6zhnty5PXcPKhqoRksC6XkenxssGHXfcjUsm9RrjOEt7Rb8InVm9hiSuGt8r9qLyfb1ij9ji\n106/+7zfHX9p/HW6nx8/64jfXzzqx4oNr7j7lMNPHzLLHc6+5pg+F7/rVGxwaa+/uycbevWP\ntbB66AlH/WPpxoeQ3PmmO9tke875vSt2HNmxS3wwpdcHjnP6NRX/PNbaVY2QGoKK73oqw2M/\npN3UTXnTq7w89AipIdCHFNl/r6pe5V7ekt9HqhohNQT6kJxvS9ZUcWlpyYoqLgUhNQwVz8MQ\nEEICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFC\nAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQOD/AYfahYIx0tuEAAAAAElFTkSu\nQmCC",
      "text/plain": [
       "Plot with title “”"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "hist(table(TTfinal$genes[TTfinal$FDRfinal < adjPvalThr & abs(TTfinal$logFC) > FCthr]), \n",
    "     breaks = 20, \n",
    "     main = '',\n",
    "     xlab=\"Pairwise comparison (#)\", \n",
    "    ylab = 'DEG (#)')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "e4e2b236-4d8e-4e45-93ca-d74256cf1da0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T17:47:44.365838Z",
     "iopub.status.busy": "2024-05-28T17:47:44.364936Z",
     "iopub.status.idle": "2024-05-28T17:47:44.380474Z",
     "shell.execute_reply": "2024-05-28T17:47:44.379355Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 6868 rows and 1 column\n",
       "                 DEG\n",
       "         <character>\n",
       "1    ENSG00000229809\n",
       "2    ENSG00000119772\n",
       "3    ENSG00000180011\n",
       "4    ENSG00000196712\n",
       "5    ENSG00000136504\n",
       "...              ...\n",
       "6864 ENSG00000037637\n",
       "6865 ENSG00000198865\n",
       "6866 ENSG00000175854\n",
       "6867 ENSG00000105426\n",
       "6868 ENSG00000198246"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DEGt <- DataFrame(DEG = DEG)\n",
    "DEGt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "94b6799b-2af6-4547-b1a1-64fd01af9e20",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-28T17:47:44.383113Z",
     "iopub.status.busy": "2024-05-28T17:47:44.382206Z",
     "iopub.status.idle": "2024-05-28T17:47:44.416524Z",
     "shell.execute_reply": "2024-05-28T17:47:44.415225Z"
    }
   },
   "outputs": [],
   "source": [
    "write.table(x = DEGt, quote = FALSE, sep = '\\t', row.names = FALSE, col.names = TRUE,\n",
    "            file = here('03_downstream_analysis/02_gene_universe_definition/results/03_DEG_healty_vs_others_list.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
